vivisect/analysis/generic/symswitchcase.py

```diff
@@ -0,0 +1,1307 @@
+            self.min_func_instr_size = vw.config.viv.analysis.symswitchcase.min_func_instr_size
+
+        except AttributeError as e:
+            logger.warning('configuration failure, using defaults: %r', e)
+            self.max_instr_count  = 10
+            self.max_cases   = 500
+            self.case_failure = 5000
+            self.min_func_instr_size = 10
+
+        self.clearCache()
+
+
+
+    def clearCache(self):
+        '''
+        While analyzing a switchcase, we end up caching several calculated items.
+        This function clears them, so we can reset the state, like to skip 
+        inappropriate paths (eg.   if x> 3 and if x == 0
+        '''
+        self.cspath = None
+        self.aspath = None
+        self.fullpath = None
+
+        self.upper = None
+        self.lower = None
+        self.count = None
+        self.baseoff = None
+        self.baseIdx = None
+
+        self.jmpsymvar = None
+
+
+    def analyze(self):
+        '''
+        Generic "analyze" function, which simply calls makeSwitch()
+        '''
+        self.makeSwitch()
+
+    
+    
+    #### primitives for switchcase analysis ####
+    def getJmpSymVar(self):
+        '''
+        returns the Simple Symbolik state of the dynamic target of the branch/jmp
+        this is the symbolik register at the start of the last codeblock.
+        '''
+        if self.jmpsymvar is not None:
+            return self.jmpsymvar
+
+        # TODO: this is not always the right operand for all architectures.  improve
+        self.jmpsymvar = self.xlate.getOperObj(self.jmpop, 0)
+        return self.jmpsymvar
+
+    def getSymTarget(self, short=True):
+        '''
+        Returns the Symbolik state of the dynamic target of the branch/jmp
+        short indicates just in the context of the last codeblock.
+        not short indicates the context at the start of the last codeblock (from funcva)
+        '''
+        jmpsymvar = self.getJmpSymVar()
+
+        cspath, aspath, fullpath = self.getSymbolikParts()
+        emu = (cspath[0], aspath[0])[short]
+        tgtsym = jmpsymvar.update(emu)
+        return tgtsym
+
+
+    def getSymIdx(self):
+        '''
+        returns the symbolik index register, and the type of object
+        '''
+        symtgt = self.getSymTarget()
+        unks = getUnknowns(symtgt)
+
+        cspath, aspath, fullpath = self.getSymbolikParts()
+
+        potentials = []
+
+        for unk in unks.get(SYMT_VAR):
+            unkvar = cspath[0].getSymVariable(unk)
+            if unkvar.isDiscrete():
+                continue
+            potentials.append((SYMT_VAR, unk))
+
+        for unk in unks.get(SYMT_MEM):
+            unkvar = cspath[0].getSymVariable(unk)
+            if unkvar.isDiscrete():
+                continue
+            potentials.append((SYMT_MEM, unk))
+
+        if not len(potentials):
+            raise v_exc.SymIdxNotFoundException(cspath, aspath)
+            
+        return potentials[0]
```

Comment:
first one should be all we need.  but there may be more than one.  this is just accounting for that possibility.
i've changed it to just return when it runs into one.
i wrote it so that we had flexibility, in case the first one wasn't good enough, but so far it seems to be acceptable.